### PR TITLE
Use tox to run Travis tests locally (deprecate runtests.sh)

### LINKED ISSUE #6190: Use tox to test similar Python scenarios as in Travis
Let's first get tox to work locally. Once we're satisfied we can follow up with another PR to switch Travis to using it.

### PR DESCRIPTION
**Pull request by shubhank-saxena**
This fixes #6190 
@armenzg this tox file is not just restricted to Python tests. I did try to integrate JS based tests as well (depends on system installation of Node, NPM, Yarn).  This will help in cleaning .travis.yml a little bit.
Do let me know if any further changes are required!

### CODE REVIEW

**Comment by armenzg** [comment]
I will look at this tomorrow.

**Comment by shubhank-saxena** [comment]
Yes, that explicit version of pip for heroku. I wasn't able to implement it. So I will reduce this just to python tests 😁

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,77 @@
+[tox]
+envlist = py37
+isolated_build = true
+skipsdist=True
```
Comment (armenzg) [inline tox.ini]: Why are these two options necessary?

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+toxworkdir={toxinidir}/.tox
```
Comment (armenzg) [inline tox.ini]: I believe the default is already this value (`toxworkdir={toxinidir}/.tox(PATH)`)
Do we need to add the value?

File: tox.ini [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+[testenv]
```
Comment (armenzg) [inline tox.ini]: Add here what runcheck.sh would run.

File: tox.ini (hunk 1 above)
Comment (shubhank-saxena) [reply tox.ini:4]: By default, Tox depends on either setup.py and pyproject.toml for installation and setup purposes. Since we want our tox env to be independent of such dependencies, theses options help to disable that

File: tox.ini (hunk 3 above)
Comment (shubhank-saxena) [reply tox.ini:8]: Okay!

File: tox.ini (hunk 2 above)
Comment (shubhank-saxena) [reply tox.ini:6]: Just an explicit declaration for our requirements and dev files to be located in a better manner. There were some issues if we were directly trying to import and execute them.

File: tox.ini (hunk 1 above)
Comment (armenzg) [reply tox.ini:4]: OK thanks.

File: tox.ini (hunk 2 above)
Comment (armenzg) [reply tox.ini:6]: Thanks for clarifying.

File: tox.ini [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+deps=
+    -r {toxinidir}/requirements/common.txt
+    -r {toxinidir}/requirements/dev.txt 
+commands_pre = 
+    pip install -r {toxinidir}/requirements/docs.txt
+    apt-get install shellcheck
```
Comment (armenzg) [inline tox.ini:14]: This will only work in a Linux system. There's another way. See comment below.

```shell
py37 run-test-pre: commands[2] | apt-get install shellcheck
ERROR: InvocationError for command could not find executable apt-get
```

File: tox.ini [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+commands =
+    pip check
+    {toxinidir}/lints/queuelint.py
+    flake8 --show-source
+    isort --check-only --diff --quiet
+    mkdocs build
+    sh -c "git grep -El '^#!/.+\b(bash|sh)\b' | xargs shellcheck"
```
Comment (armenzg) [inline tox.ini:21]: Run this instead: `pre-commit run --all-files shellcheck`

This removes the need to install `shellcheck`.

File: tox.ini [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+[testenv:outer]
```
Comment (armenzg) [inline tox.ini:23]: Call this one `virtualenv` and the next section `docker`.

File: tox.ini [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+deps = 
+    -r {toxinidir}/requirements/common.txt
+    -r {toxinidir}/requirements/dev.txt
+    docker-compose
```
Comment (armenzg) [inline tox.ini:27]: Please validate that each environment works. None of them currently work for me.

File: tox.ini [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    {toxinidir}/runchecks.sh
+    {toxinidir}/manage.py check
+    SITE_URL=https://treeherder.dev TREEHERDER_DEBUG=False {toxinidir}/manage.py check --deploy --fail-level WARNING
+    pytest tests/ --ignore=tests/selenium --ignore=tests/extract
+    docker-compose down
+
+[testenv:inner]
+deps = 
+    codecov
+    docker-compose
```
Comment (armenzg) [inline tox.ini:40]: This adds the `docker-compose` Python package which we don't need.

File: tox.ini (hunk 7 above)
Comment (shubhank-saxena) [reply tox.ini:27]: I think I need to merge slimmer branch and then I think they should work. Let me do that!

File: tox.ini [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+commands_pre = 
+    docker-compose build
+commands =
+    docker-compose run backend bash -c "pytest --cov --cov-report=xml tests/ --runslow --ignore=tests/selenium"
+    codecov -f coverage.xml
```
Comment (armenzg) [inline tox.ini:45]: We should leave the `codecov` commands directly in Travis since most of us we will not be attempting to upload the artifact to the Codecov service. Only Travis will be doing that.

File: tox.ini [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+[testenv:selenium]
+deps = 
+    codecov
+    docker-compose
+whitelist_externals =
+    yarn
```
Comment (armenzg) [inline tox.ini:52]: You should whitelist `docker-compose` in all sections where it's used.

File: tox.ini (hunk 8 above)
Comment (shubhank-saxena) [reply tox.ini:40]: Okay. Will whitelist it

File: tox.ini (hunk 10 above)
Comment (shubhank-saxena) [reply tox.ini:52]: Understood. I blankly replicated everything. It's better to leave installations to Travis

**Comment by shubhank-saxena** [comment]
Oh no! What have I done!

**Comment by SuyashSalampuria** [comment]
Actually, you did not rebase while pulling stuff from other branches

**Comment by shubhank-saxena** [comment]
> Actually, you did not rebase while pulling stuff from other branches

I actually did do git fetch upstream. Will see it again! Thank you!

**Comment by shubhank-saxena** [comment]
@armenzg waiting for slimmer PR to merge. I did try rebasing, but it is creating a lot of issues and iirelevant changes to this PR being added. Will merge master and continue. As far I have tested, they were working fine on the previous editions of the .txt files

**Comment by armenzg** [comment]
I'm more familiar with this PR. I will spare camd from it.

**Comment by shubhank-saxena** [comment]
> I'm more familiar with this PR. I will spare camd from it.

Apologies. It happened while I was rebasing. Did reverted the commits. But wasn't able to un-assign the reviews

**Comment by shubhank-saxena** [comment]
@armenzg I have tested all the commands locally. They are working fine. But for this particular task that failed in Travis, the error looks specific for Travis
```django.db.utils.OperationalError: (2013, "Lost connection to MySQL server at 'reading initial communication packet', system error: 0")```
Checking the workaround

File: .travis.yml [hunk 11]
```diff
@@ -60,33 +60,20 @@ jobs:
       python: '3.7'
       cache: pip
       install:
-        # Initialize services required to run tests
-        - docker-compose up --detach mysql redis rabbitmq
```
Comment (armenzg) [inline .travis.yml:64]: The command "manage.py check" is failing because there's no database running. Add this back.

File: .travis.yml [hunk 12]
```diff
@@ -60,33 +60,20 @@ jobs:
...
-        # Several security features in settings.py (eg setting HSTS headers) are conditional on
-        # 'https://' being in the site URL. In addition, we override the test environment's debug
-        # value so the tests pass. The real environment variable will be checked during deployment.
-        - SITE_URL=https://treeherder.dev TREEHERDER_DEBUG=False ./manage.py check --deploy --fail-level WARNING
-        # Exercising running tests outside of Docker
-        - pytest tests/ --ignore=tests/selenium --ignore=tests/extract
-        # This is to deal with running the containers with --detached
-        - docker-compose down
+        - tox
+        - tox -e virtualenv
```
Comment (armenzg) [inline .travis.yml:66]: I am thinking that running two tox environments will cause the installation of packages twice.
Let's see if we can fold the two together.

File: tox.ini [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+commands =
+    pip check
+    {toxinidir}/lints/queuelint.py
+    flake8 --show-source
+    isort --check-only --diff --quiet
+    mkdocs build
+    pre-commit run --all-files shellcheck
+    pre-commit run --all-files markdownlint
+
+[testenv:virtualenv]
```
Comment (armenzg) [inline tox.ini:22]: Merge the `virtualenv` environment into the default one.

Comment (shubhank-saxena) [reply tox.ini:22]: I wanted to keep it different so that we have seperate of runchecks.sh and rest everything different. But I can merge them!

**Comment by armenzg** [comment]
Iterate few times until you get Travis working right.
For short questions feel free to ask directly on Riot after you've tried few things. That way you don't wait few days until I review again.

**Comment by shubhank-saxena** [comment]
@armenzg done!

**Comment by shubhank-saxena** [comment]
> From my point of view this is working well enough.
> Please follow up on any comments from Kyle.
> 
> Note: I always forget to run this command `docker-compose up --detach mysql redis rabbitmq` before running `tox`. Is it possible for `tox` to take care of starting those services up? Thanks!

@armenzg I did try to whitelist docker-compose and shifted this script, but it was failing on Travis

**Comment by armenzg** [comment]
> > From my point of view this is working well enough.
> > Please follow up on any comments from Kyle.
> > Note: I always forget to run this command `docker-compose up --detach mysql redis rabbitmq` before running `tox`. Is it possible for `tox` to take care of starting those services up? Thanks!
> 
> @armenzg I did try to whitelist docker-compose and shifted this script, but it was failing on Travis

What was the issue?

If you want we can defer solving the issue to another PR since I would rather see this landed first. Meanwhile make sure to add a note that the command needs to be run first (unless you figure it out).

**Comment by shubhank-saxena** [comment]
@armenzg done!

File: tox.ini [hunk 14]
```diff
@@ -23,6 +25,8 @@ commands =
     sh -c "SITE_URL=https://treeherder.dev TREEHERDER_DEBUG=False ./manage.py check --deploy --fail-level WARNING"
     # Exercising running tests outside of Docker
     pytest tests/ --ignore=tests/selenium --ignore=tests/extract
+    # This is to deal with running the containers with --detached
+    docker-compose down
```
Comment (armenzg) [inline tox.ini:29]: I believe it is better to use with [commands_post](https://tox.readthedocs.io/en/latest/config.html#conf-commands_post)

File: requirements/dev.txt [hunk 15]
```diff
@@ -12,10 +12,6 @@ asgiref==3.2.7 \
     --hash=sha256:8036f90603c54e93521e5777b2b9a39ba1bad05773fcf2d208f0299d1df58ce5 \
     --hash=sha256:9ca8b952a0a9afa61d30aa6d3d9b570bb3fd6bafcf7ec9e6bed43b936133db1c \
     # via django
-atomicwrites==1.3.0 \
```
Comment (armenzg) [inline requirements/dev.txt:15]: Are these changes intentional?

**Comment by shubhank-saxena** [comment]
@klahnakoski do let me know if any further changes are required

**Comment by klahnakoski** [comment]
@shubhank-saxena  I have nothing to say. I do not know why this change is happening; is there a bug number?  It looks like it is linux/mac specific, so it does not affect local windows development.

**Comment by shubhank-saxena** [comment]
Actually this PR aims to remove need of runchecks.sh, so that local testing on windows can be done in a better way. Also, this PR has replicated all of Python based tests in Travis, so that exact tests can be performed locally before pushing.

File: tox.ini [hunk 16]
```diff
@@ -0,0 +1,51 @@
...
+    pip install -r {toxinidir}/requirements/docs.txt
+    {toxinidir}/manage.py check
+commands =
+    pip check
+    {toxinidir}/lints/queuelint.py
+    flake8 --show-source
+    mkdocs build
+    pre-commit run --all-files shellcheck
+    pre-commit run --all-files markdownlint
+    sh -c "SITE_URL=https://treeherder.dev TREEHERDER_DEBUG=False ./manage.py check --deploy --fail-level WARNING"
```
Comment (klahnakoski) [inline tox.ini:25]: @shubhank-saxena are these lines running in a docker container?  this will not work on Windows

Comment (armenzg) [reply tox.ini:25]: You can probably drop `sh -c` and it will still work.

File: requirements/dev.txt (hunk 15 above)
Comment (armenzg) [reply requirements/dev.txt:15]: Please revert the `dev.txt` changes.

File: tox.ini (hunk 16 above)
Comment (shubhank-saxena) [reply tox.ini:25]: > Reverting `dev.txt` changes and removing the use of `sh` should be sufficient to land this. Thanks!

The issue of using sh -c is that tox wasn't handling env setting properly

Comment (armenzg) [reply tox.ini:25]: OK; handle the `dev.txt` and we can look into it later.

**Comment by shubhank-saxena** [comment]
@armenzg done!
